quich.1
.\" Manpage for quich.

.TH QUICH 1 2021-09-23 "1.0" quich

.SH NAME
quich \- Just an advanced terminal calculator.

.SH SYNOPSIS
.B quich
[\fBOPTIONS\fR]

.SH DESCRIPTION
.B Quich
is a compact, fast, powerful and useful calculator for your terminal with numerous features, supporting Linux, Mac and Windows, written in ANSI C.
You can enter in interactive mode by calling Quich in your terminal without an operation.

.SS Functions
.B Syntax:
\fIfunc(operand)\fR
.TP
.BR sqrt
Square root
.TP
.BR abs
Absolute value (positive)
.TP
.BR log
Natural logarithm
.TP
.BR sin ", " cos ", " tan
Sine, Cosine and Tangent trigonometric functions
.TP
.BR asin ", " acos ", " atan
Arc sine, Arc cosine and Arc tangent trigonometric functions
.TP
.BR rand
Random number between 0 and 1
.TP
.BR round
Round to the nearest integer value
.TP
.BR floor
Round down
.TP
.BR ceil
Round up

.SH OPTIONS
.TP
.BR \-d ", " \-\-degree
Manage the given angles in degrees
.TP
.BR \-f ", " \-\-format\ \fI[%s]\fR
The format to display the result
.TP
.BR \-h ", " \-\-help
Get help and information
.TP
.BR \-i ", " \-\-interactive
Force interactive mode
.TP
.BR \-p ", " \-\-precision\ \fI[%i]\fR
The number of decimals to be used for the internal numbers
.TP
.BR \-r ", " \-\-round\ \fI[%i]\fR
The number of decimals to round the result
.TP
.BR \-t ", " \-\-thousands
Display the result with thousands separators
.TP
.BR \-vvv ", " \-\-verbose
Display the result with details
.TP
.BR \-v ", " \-\-version
Show the application version

.SH EXAMPLES

    quich 5+3
    8

    quich "a=20;a+1"
    21

    quich "5+(cos( 2 )-2)^2"
    10.8377655357568

    quich "5+(cos( 2 )-2)^2" -p 2
    10.86

    quich 1234567+1 -t
    1,234,568

    quich 1gb+1mb
    1049600

    quich 12345 -f '%.1g'
    1e+04

    quich 5+PI -vvv
    Tokens > '5' '+' 'PI'
    Posfix > 5 PI +
    Result > 8.14159265358979


CHANGELOG.md
# Release Notes

## v4.0.0 (2021-07-07)

### Changed

- REPL functionality.

### Fixed

- Operations validation.

## v3.2.1 (2021-04-21)

### Fixed

- Bug in lexer related with signed numbers.

- Segmentation fault in operations with an invalid number of opening/closing parentheses.

## v3.2.0 (2021-04-20)

### Added

- Warning message for wrong number of parenthesis.

## v3.1.1 (2021-04-20)

### Fixed

- Memory leaks.

## v3.1.0 (2021-03-04)

### Added

- `GR` constant.

## v3.0.0 (2020-09-02)

### Changed

- Variables work by value instead of by reference.

- Minor performance improvements.

## v2.6.2 (2020-08-05)

### Fixed

- Memory leaks.

## v2.6.1 (2020-06-10)

### Fixed

- Bug related to variable declaration.

## v2.5.0 (2020-06-08)

### Added

- `rand` function that returns a random number between 0 and 1.

- `format` flag for formatting the result.

- `G` constant.

## v2.4.0 (2020-06-06)

### Added

- `thousands` flags to display thousands separators.

## v2.3.0 (2020-05-30)

### Added

- Support for numbers with commas (,).

- Data measurement units (mb, gb, tb and pt).

## v2.2.0 (2020-03-31)

### Added

- `asin`, `acos` and `atan` functions.

- `interactive` flag to force interactive mode.

### Fixed

- Memory error message.

## v2.1.0 (2020-03-23)

### Added

- Validation for division by zero.

- Question mark for every undefined token in verbose mode.

### Fixed

- Bug related to operands and parenthesis.

## v2.0.0 (2020-03-22)

### Added

- Flags in interactive mode.

- `degree` and `round` flags.

- `PI` and `E` constants.

### Fixed

- Bug related to negative exponent numbers.

### Removed

- The `%` operator.

## v1.1.2 (2020-03-04)

### Fixed

- Removed extra whitespaces.

## v1.0.0 (2020-03-03)

- Initial release.


src/quich.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include "../lib/linenoise.h"
#include "helper.h"
#include "lexer.h"
#include "parser.h"
#include "variable.h"
#include "quich.h"

/**
 * Verbose mode or not.
 */
bool verbose = false;

/**
 * Number of flags used.
 */
int flags_quantity = 0;

/**
 * Interactive mode or not.
 */
bool interactive_mode = false;

/**
 * Using thousands separator or not.
 */
bool thousands_sep = false;

/**
 * Format to display the results.
 */
char *format = NULL;


static void printVerbose(struct list *tokens, struct list *output)
{
    struct token *node = tokens->first;

    /* Tokens */
    printf("Tokens > ");
    while (node != NULL) {
        printf(isValid(node) ? "'%s' " : "'%s'? ", node->value);
        node = node->next;
    }

    node = output->first;

    /* Postfix operation */
    printf("\nPosfix > ");
    while (node != NULL) {
        printf("%s ", node->value);
        node = node->next;
    }

    printf("\nResult > ");
}


static void printResult(char *func)
{
    struct list *tokens, *output;
    char *result;

    initList(&tokens);
    initList(&output);

    result = getResult(func, tokens, output);

    if (!isEmpty(format)) {
        snprintf(result, BUFFER, format, strToDouble(result));
    }

    if (verbose) {
        printVerbose(tokens, output);
    }

    if (thousands_sep) {
        addThousandsSep(result);
    }

    if (result != NULL) {
        printf("%s\n", result);
        free(result);
    } else if (verbose) {
        printf(MSG_DEFINITION);
    }

    printWarnings(output);
    freeList(tokens);
    freeList(output);
}


static void printAll(char *func)
{
    char *statements = strtok(func, STMT_SEPARATOR);

    while (statements != NULL) {
        printResult(statements);
        statements = strtok(NULL, STMT_SEPARATOR);

        if (statements != NULL && verbose) {
            printf("\n");
        }
    }
}


/**
 * Returns true if the program flow must be stopped,
 * false otherwise.
 */
static bool mapArgs(int argc, char *argv[])
{
    int i;

    for (i = 0; i < argc; i++) {
        /* Angles in degree */
        if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--degree")) {
            degree = true;
            flags_quantity++;
        }

        /* Verbose mode */
        if (!strcmp(argv[i], "-vvv") || !strcmp(argv[i], "--verbose")) {
            verbose = true;
            flags_quantity++;
        }

        /* Interactive */
        if (!strcmp(argv[i], "-i") || !strcmp(argv[i], "--interactive")) {
            interactive_mode = true;
            flags_quantity++;
        }

        /* Version */
        if (!strcmp(argv[i], "-v") || !strcmp(argv[i], "--version")) {
            printf(MSG_VERSION);
            return true;
        }

        /* Help */
        if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
            printf(MSG_HELP);
            return true;
        }

        /* Thousands separator */
        if (!strcmp(argv[i], "-t") || !strcmp(argv[i], "--thousands")) {
            thousands_sep = true;
            flags_quantity++;
        }

        /* The flags below work with values */
        if (i+1 >= argc) {
            return false;
        }

        /* Result format */
        if (!strcmp(argv[i], "-f") || !strcmp(argv[i], "--format")) {
            format = argv[++i];
            flags_quantity += 2;
        }

        /* Precision */
        if (!strcmp(argv[i], "-p") || !strcmp(argv[i], "--precision")) {
            precision = (int)strToDouble(argv[++i]);
            flags_quantity += 2;
        }

        /* Result precision */
        if (!strcmp(argv[i], "-r") || !strcmp(argv[i], "--round")) {
            result_precision = (int)strToDouble(argv[++i]);
            flags_quantity += 2;
        }
    }

    return false;
}


static void interactive(void)
{
    char *line = NULL;
    printf(MSG_INIT);

    while ((line = linenoise("> ")) != NULL && strcmp(line, EXIT_COMMAND) != 0) {
        linenoiseHistoryAdd(line);
        printAll(line);
        linenoiseFree(line);
    }

    linenoiseFree(line);
    printf(MSG_BYE);
}


static void addPredefValues(void)
{
    addVariable("PI", MATH_PI);
    addVariable("E", MATH_E);
    addVariable("G", MATH_G);
    addVariable("GR", MATH_GR);
}


int main(int argc, char* argv[])
{
    if (mapArgs(argc, argv)) {
        return 0;
    }

    addPredefValues();
    if (interactive_mode || flags_quantity >= argc - 1) {
        interactive();
    } else {
        printAll(argv[1]);
    }

    freeVariables();

    return 0;
}
